src/v/datalake/coordinator/tests/state_machine_test.cc

```diff
@@ -0,0 +1,195 @@
+            const auto& p_state = it->second.pid_to_pending_files;
+            auto p_it = p_state.find(tp.partition);
+            if (p_it == p_state.end()) {
+                result.emplace_back(std::nullopt);
+                continue;
+            }
+            auto& pending_files = p_it->second.pending_entries;
+            if (!pending_files.empty()) {
+                result.emplace_back(pending_files.back().data.last_offset);
+            } else {
+                result.push_back(p_it->second.last_committed);
+            }
+        }
+        return result;
+    }
+
+    model::topic_partition random_tp() const {
+        return {
+          model::topic{"test"},
+          model::partition_id(
+            random_generators::get_int<int32_t>(0, max_partitions - 1))};
+    }
+
+    static constexpr int32_t max_partitions = 5;
+    model::topic_partition tp{model::topic{"test"}, model::partition_id{0}};
+    datalake::coordinator::simple_file_committer file_committer;
+    absl::flat_hash_map<raft::vnode, coordinator> coordinators;
+};
+
+TEST_F_CORO(coordinator_stm_fixture, test_snapshots) {
+    co_await initialize();
+    co_await wait_for_leader(5s);
+
+    // populate some data until the state machine is snapshotted
+    // a few times
+    constexpr auto max_snapshots = 5;
+    auto completed_snapshots = 0;
+    auto prev_snapshot_offset = last_snapshot_offset();
+    while (completed_snapshots != max_snapshots) {
+        // mock a translator
+        auto add_files_result = co_await retry_with_leader_coordinator(
+          [&, this](coordinator& coordinator) mutable {
+              auto tp = random_tp();
+              return coordinator->sync_get_last_added_offset(tp).then(
+                [&, tp](auto result) {
+                    if (!result) {
+                        return ss::make_ready_future<bool>(false);
+                    }
+                    auto last_committed_offset = kafka::offset_cast(
+                      result.value().value_or(kafka::offset{-1}));
+                    std::vector<std::pair<int64_t, int64_t>> offset_pairs;
+                    offset_pairs.reserve(5);
+                    auto next_offset = last_committed_offset() + 1;
+                    for (int i = 0; i < 5; i++) {
+                        offset_pairs.emplace_back(next_offset, next_offset + 5);
+                        next_offset = next_offset + 6;
+                    }
+                    return coordinator
+                      ->sync_add_files(
+                        tp,
+                        datalake::coordinator::make_pending_files(offset_pairs))
+                      .then([](auto result) {
+                          return ss::make_ready_future<bool>(
+                            result.has_value());
+                      });
+                });
+          });
+        ASSERT_TRUE_CORO(add_files_result) << "Timed out waiting to add files";
+        auto snapshot_offset = last_snapshot_offset();
```

Comment:
Is it important that this is only checking one node, that may not be the leader? Wondering if there's a pathological case where it causes the below check on the new node's start offset to fail?